src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ProducerConsumerStream.cs

```diff
@@ -0,0 +1,350 @@
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+            if (offset < 0 || offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
+            }
+            if (count < 0 || count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            Contract.EndContractBlock();
+
+            if (_disposed)
+            {
+                return 0;
+            }
+
+            _buffer.SetResult(new WriteBufferWrapper(buffer, offset, count));
+            int totalBytesRead = 0;
+            do
+            {
+                int bytesRead = _dataAvail.Task.Result;
+                _dataAvail = new TaskCompletionSource<int>();
+                totalBytesRead += bytesRead;
+                count -= bytesRead;
+                if (bytesRead == 0)
+                {
+                    break;
+                }
+            } while (count > 0);
+            return totalBytesRead;
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+            if (offset < 0 || offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
+            }
+            if (count < 0 || count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            Contract.EndContractBlock();
+
+            if (_disposed)
+            {
+                throw new ObjectDisposedException("ProducerConsumerStream");
+            }
+
+            while (count > 0)
+            {
+                if (_currentBuffer == WriteBufferWrapper.EmptyContainer)
+                {
+                    _currentBuffer = _buffer.Task.Result;
+                    _buffer = new TaskCompletionSource<WriteBufferWrapper>();
+                }
+
+                int bytesWritten = _currentBuffer.Write(buffer, offset, count);
+                count -= bytesWritten;
+                offset += bytesWritten;
+                if (_currentBuffer.Count == 0)
+                {
+                    _currentBuffer = WriteBufferWrapper.EmptyContainer;
+                }
+
+                _dataAvail.TrySetResult(bytesWritten);
+            }
+        }
+
+        public override bool CanRead
+        {
+            get { return !_disposed; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return false; }
+        }
+
+        public override bool CanWrite
+        {
+            get { return !_disposed; }
+        }
+
+        public override long Length
+        {
+            get { throw new NotImplementedException(); }
```

Comment:
As per response to Jason, `NotSupportedException` is the correct exception so I'm changing it to that.